src/System.IO.Pipelines/MemoryPoolBlock.cs

```diff
@@ -86,20 +86,15 @@ internal void Initialize()
         public override string ToString()
         {
             var builder = new StringBuilder();
-            var data = Memory.Span;
-
-            for (int i = 0; i < data.Length; i++)
-            {
-                builder.Append((char)data[i]);
-            }
+            SpanExtensions.AppendAsLiteral(Memory.Span, builder);
             return builder.ToString();
         }
 
         protected override void Dispose(bool disposing)
         {
-            Pool.Return(this);
-
             base.Dispose(disposing);
+
+            Pool.Return(this);
```

Comment:
We need to dispose first then return because we can get into situation where we dispose newly leased block.